feat: getPath with carScope

olizilla opened the pull request:
add getPath method as a generator that returns blocks for the targeted dag and all blocks traversed while resolving a cid+path string

supports carScope to specify what blocks to return for the resolved dag
- `'all'`: return the entire dag starting at path. (default)
- `'block'`: return the block identified by the path.
- `'file'`: Mimic gateway semantics: Return All blocks for a multi-block file or just enough blocks to enumerate a dir/map but not the dir contents.

see: https://github.com/web3-storage/freeway/issues/33
see: https://github.com/web3-storage/freeway/issues/34
see: https://github.com/ipfs/specs/pull/402

TODO:
- [x] find out how to identify the boundaries of a unixfs hamt 

...unixfs-exporter seems to define it as "not having an empty or null Link.Name after the first 2 chars are stripped, which seems loose... what happens if the actual dir listing has 2 char long link names? see: https://github.com/ipfs/js-ipfs-unixfs/blob/e853049bd63d6773442e1540ae49b6a443ca8672/packages/ipfs-unixfs-exporter/src/resolvers/unixfs-v1/content/hamt-sharded-directory.ts#L20-L42

License: MIT

olizilla commented:
@alanshaw plz to sanity check. also if you can point me at the right way to detect the boundary of the current hamt then that would save me some staring at things.

olizilla commented on package.json:
```diff
@@ -49,8 +49,7 @@
     "ipfs-unixfs": "^11.0.0",
     "miniswap": "^2.0.0",
-    "standard": "^17.0.0",
-    "uint8arrays": "^3.0.0"
```
turns out the swiss-army-knife has `toString` and `fromString` under `multiformats/bytes`! who knew!

alanshaw commented on index.js:
```diff
@@ -66,14 +66,14 @@ export class Dagula {
-    let cids = [cid]
-    while (true) {
+    let cids = Array.isArray(cid) ? cid : [cid]
+    while (cids.length > 0) {
```
Can probably remove `if (!nextCids.length) break` later if doing the check up here...

alanshaw commented on index.d.ts:
```diff
@@ -61,7 +75,7 @@ export declare class Dagula implements IDagula {
    * Emit nodes for all path segements and get UnixFS files and directories
    */
-  walkUnixfsPath (path: CID|string, options?: AbortOptions): Promise<UnixFSEntry>
+  walkUnixfsPath (path: CID|string, options?: AbortOptions): AsyncIterableIterator<UnixFSEntry & Block>
```
Is `UnixFSEntry` not already a `Block`?

olizilla replied:
No, it lacks the all important `bytes`

alanshaw reviewed (approved):
LGTM

alanshaw commented:
> ...unixfs-exporter seems to define it as "not having an empty or null Link.Name after the first 2 chars are stripped, which seems loose... what happens if the actual dir listing has 2 char long link names?

You just strip the first two characters, and if the result is empty then it's a HAMT node, and if it is not empty it is not.

alanshaw commented on index.js:
```diff
@@ -88,16 +91,52 @@ export class Dagula {
+    const carScope = options.carScope ?? 'all'
+    /** @type {import('ipfs-unixfs-exporter').UnixFSEntry} */
+    let base
+    for await (const item of this.walkUnixfsPath(cidPath, { signal: options.signal })) {
+      base = item
+      yield item
+    }
```
Apologies I should have caught that in the review!

olizilla commented:
@alanshaw ok the bits i missed first time around are now fixed, plz to eyeball

1. as you pointed out `walkPath` emits a single entry even for multi-block structures: fixed by passing in a custom blockstore that caches the traversed blocks so we can emit them all.
2. i was using the `multiformat/block.links` function in my linkFilter, but that only understands ipld flavour link parsing, so dag-pb nodes link names are `Links/<index>/Hash` regardless of a `Name` field being present, so my hamtSearch wasn't working: fixed by special casing dagPB, which it would need to be anyway.

I've gone this root to ride on the work in unixfs-exporter for today, and get it working, hopefully without introducing any extra edge cases. I plan to remove this logic from here next week and put it in to https://github.com/web3-storage/ipfs-path which will become a block-centric rather than unixfsEntry-centric lib just for resolving ipfs paths... but first i want to land and ship verifiable paths and carScope to our gateway.

alanshaw reviewed (approved):
LGTM
